src/compiler/program.ts

```diff
@@ -553,9 +553,39 @@ namespace ts {
         return getModeForUsageLocation(file, getModuleNameStringLiteralAt(file, index));
     }
 
+    /* @internal */
+    export function isExclusivelyTypeOnlyImportOrExport(decl: ImportDeclaration | ExportDeclaration) {
+        if (isExportDeclaration(decl)) {
+            return decl.isTypeOnly;
+        }
+        if (decl.importClause?.isTypeOnly) {
+            return true;
+        }
+        const bindings = decl.importClause?.namedBindings;
+        if (bindings && isNamedImports(bindings) && length(bindings.elements) && every(bindings.elements, b => b.isTypeOnly)) {
+            return true;
+        }

```

Comment:
This shouldn’t be considered here—at the very least, you need to check that `compilerOptions.importsNotUsedAsValues` is not `preserve` or `error`, but I think the more coherent solution is that only `import type ...` is allowed to specify an override. A regular import whose specifiers happen to all be marked type-only is really no different than a regular import who happens to import only types. I want to reinforce the mental model that `import type ...` → no emit (therefore can have special behavior) and `import { type Foo }` → `import {}`. If the latter gets elided from JS emit, that’s a follow-on effect of our unusually aggressive import elision; it doesn’t mean the source import was special.